misc/example/m.h

```diff
@@ -3,92 +3,55 @@
 #include "out.h"
 
 #include <array>
+#include <cassert>
 #include <hpx/hpx.hpp>
 #include <iostream>
 #include <memory>
 
 template <class El>
-class Tile : public Tile<const El> {
-public:
-  Tile(El* ptr) : Tile<const El>(ptr), ptr_(ptr) {}
-
-  Tile(const Tile&) = delete;
-  Tile(Tile&& rhs) : Tile<const El>(std::move(rhs)), ptr_(rhs.ptr_) {
-    // std::cout << "Move" << std::endl;
-    rhs.ptr_ = nullptr;
-  }
-
-  El& operator()() const {
-    return *ptr_;
-  }
-
-private:
-  El* ptr_;
-};
-
-template <class El>
-class Tile<const El> {
-public:
-  Tile(const El* ptr) : ptr_(ptr) {}
-
-  Tile(const Tile&) = delete;
-  Tile(Tile&& rhs) : ptr_(rhs.ptr_) {
-    // std::cout << "Move" << std::endl;
-    rhs.ptr_ = nullptr;
-  }
-
-  const El& operator()() const {
-    return *ptr_;
-  }
-
-private:
-  const El* ptr_;
-};
-
-template <class El>
-class Wrapper : public Wrapper<const El> {
-  using Type = Tile<El>;
-  using Wrapper<const El>::t_;
+class Tile {
+  using ncEl = std::remove_const_t<El>;
+  using cEl = std::add_const_t<El>;
+  friend Tile<ncEl>;
+  friend Tile<cEl>;
 
 public:
-  Wrapper(Type&& t, hpx::promise<Type>&& p) : Wrapper<const El>(std::move(t), std::move(p)) {}
+  Tile(El* ptr) : ptr_(ptr), p_(nullptr) {}
 
-  Wrapper(const Wrapper&) = delete;
-  Wrapper(Wrapper&& rhs) = default;
+  template <class T = El>
+  Tile(std::enable_if_t<std::is_same<T, El>::value && !std::is_const<T>::value, El*> ptr,
+       hpx::promise<Tile<El>>&& p)
+      : ptr_(ptr), p_(std::make_unique<hpx::promise<Tile>>(std::move(p))) {}
 
-  const Type& get() const {
-    return t_;
-  }
-};
-
-template <class El>
-class Wrapper<const El> {
-  friend Wrapper<El>;
-  using Type = Tile<El>;
-  using ConstType = Tile<const El>;
+  Tile(const Tile&) = delete;
+  Tile(Tile&& rhs) = default;
 
-public:
-  Wrapper(Type&& t, hpx::promise<Type>&& p) : t_(std::move(t)), p_(std::move(p)), valid_(true) {}
+  template <class T = El>
+  Tile(std::enable_if_t<std::is_same<T, El>::value && std::is_const<T>::value, Tile<ncEl>&&> rhs)
+      : ptr_(rhs.ptr_), p_(std::move(rhs.p_)) {}
 
-  ~Wrapper() {
-    if (valid_) {
-      p_.set_value(std::move(t_));
+  ~Tile() {
+    if (p_) {
+      auto p = std::move(p_);
+      p->set_value(Tile<ncEl>(ptr_));
     }
   }
 
-  Wrapper(const Wrapper&) = delete;
-  Wrapper(Wrapper&& rhs) : t_(std::move(rhs.t_)), p_(std::move(rhs.p_)), valid_(true) {
-    rhs.valid_ = false;
+  template <class T = El>
+  std::enable_if_t<std::is_same<T, El>::value && !std::is_const<T>::value, Tile&> setPromise(
```

Comment:
Do we really want to expose this in the public interface?

I think that `setPromise` should be called with caution, so I would go for managing it internally without allowing anyone who has a `Tile` to be able to call it: if it is called multiple times on the same `Tile` it is a problem.